odoo/tests/common.py

```diff
@@ -2060,17 +2038,20 @@ def setUp(self):
         # setup an url opener helper
         self.opener = Opener(self)
         self.http_key_sequence = itertools.count()
-        # we need to allow requests during pdf rendering.
-        old_run_wkhtmltopdf = ir_actions_report._run_wkhtmltopdf
-
-        def _patched_run_wkhtmltopdf(args):
-            with patch.object(self, 'http_request_key', 'wkhtmltopdf'), release_test_lock():
-                args = ['--cookie', TEST_CURSOR_COOKIE_NAME, 'wkhtmltopdf', *args]
-                return old_run_wkhtmltopdf(args)
+        self.is_pdf_rendering_allowed = False
 
-        self.startPatcher(
-            patch.object(ir_actions_report, '_run_wkhtmltopdf', _patched_run_wkhtmltopdf),
-        )
+    @contextmanager
+    def allow_pdf_render(self):
+        """
+        Allows wkhtmltopdf to send requests to the backend.
+        Enters registry mode if necessary.
+        """
+        with ExitStack() as stack:
+            self.is_pdf_rendering_allowed = True
+            if not type(self)._registry_patched:
+                stack.enter_context(self.enter_registry_test_mode())
+            yield
+            self.is_pdf_rendering_allowed = False
```

Comment:
If the test fails, you still need to unset the flag on self:

```suggestion
        assert not self.is_pdf_rendering_allowed, "reentry not implemented"
        try:
            self.is_pdf_rendering_allowed = True
            with self.enter_registry_test_model() if not type(self)._registry_patched else null_context():
                yield
        finally:
            self.is_pdf_rendering_allowed = False
```